[aspnet] Initial implementation of ASP.NET 5 server

jimschubert opened the pull request:
See Issue #1469

Things to work on:

Gaps:

* Missing swagger definition functionality:
  - defaultResponse
  - examples
  - auth
  - consumes
  - produces
  - nickname
  - externalDocs
  - imports
  - security
  - schema
* Resolve allParams/hasMore issue with headerParams
* Create functional test project stub
* Resolve all issues with value type return values

In this commit:

* Initial cross-platform ASP.NET 5 API server
* Hook up swagger gen via Swashbuckle and xml comment
* Build script (*nix) in project root
* Dockerfile for container deployment

wing328 commented:
@jimschubert thank you :beers: 

I'll review and let you know if I've any questions.

jimschubert commented:
@wing328 Thanks. It generates a pretty complete server. Nothing snazzy like services or DI. Just pushed a second commit to fix a bad merge of the Haskell Servant example in README.md when I rebased.

I did have to add a hack for operation paths. Something like `/pet/{petId}?testing_byte_array=true` is an invalid identifier for [RouteAttribute](https://msdn.microsoft.com/en-us/library/system.web.http.routeattribute(v=vs.118).aspx). I just replaced the `?` with `/` so the user can build and modify as needed.

wing328 commented on modules/swagger-codegen/src/main/resources/aspnet5/model.mustache:
```diff
@@ -0,0 +1,134 @@
+    /// <summary>
+    /// {{description}}
+    /// </summary>
+    public class {{classname}} : {{#parent}}{{{parent}}}, {{/parent}} IEquatable<{{classname}}>
```
[OPTIONAL] we just added `partial` to the model class. You might want to do the same.

jimschubert replied:
I can add that. After we recently regenerated the C# client at work, I was thinking about how nice it would be if this was had the partial keyword.

I'll look at the client code, but are there any plans or best practices for partial methods? Or just the class for now?

wing328 replied:
To me, adding partial to model classes is just a starting point. If there're demands/use cases for adding partial to methods or other classes, we'll definitely consider it.

wing328 commented on README.md:
````diff
@@ -587,6 +588,15 @@ java -jar modules/swagger-codegen-cli/target/swagger-codegen-cli.jar generate \
+```
+java -jar modules/swagger-codegen-cli/target/swagger-codegen-cli.jar generate \
+  -i http://petstore.swagger.io/v2/swagger.json \
+  -l aspnet5-webapi \
````
Should this be `aspnet5` (not `aspnet5-webapi`)?

jimschubert replied:
Yeah, thanks for catching this. I had to abort my first rebase in which I changed this, and missed it the second rebase.

jimschubert replied:
I'll leave the header as `ASP.NET 5 Web API`. Technically, MVC and Web API are being merged and this is why I renamed from `aspnet5-webapi` to just `aspnet5`. I think the header will make it clear for now that there aren't any server-side views, and that the generator focuses only on the api.

wing328 commented:
[OPTIONAL] you may also want add `\bin\windows\aspnet5-petstore-server.bat` as there may be more developers using it on the Windows platform. (ref: https://github.com/swagger-api/swagger-codegen/blob/master/bin/windows/csharp-petstore.bat)

jimschubert commented:
@wing328 I made those changes. Let me know if there's anything else you find.

wing328 commented:
Looks good to me.
